autoprecompiles/src/low_degree_bus_interaction_optimizer.rs

```diff
@@ -0,0 +1,562 @@
+use itertools::Itertools;
+use powdr_constraint_solver::constraint_system::{
+    BusInteraction, BusInteractionHandler, ConstraintSystem,
+};
+use powdr_constraint_solver::grouped_expression::{GroupedExpression, NoRangeConstraints};
+use powdr_constraint_solver::inliner::DegreeBound;
+use powdr_constraint_solver::range_constraint::RangeConstraint;
+use powdr_constraint_solver::runtime_constant::RuntimeConstant;
+use powdr_constraint_solver::solver::Solver;
+use powdr_number::FieldElement;
+use powdr_number::LargeInt;
+use std::collections::BTreeMap;
+use std::fmt::Display;
+use std::hash::Hash;
+use std::marker::PhantomData;
+
+use crate::constraint_optimizer::IsBusStateful;
+use crate::range_constraint_optimizer::{RangeConstraintHandler, RangeConstraints};
+
+/// An optimizer that replaces some stateless bus interactions (a.k.a. lookups)
+/// by low-degree algebraic constraints.
+pub struct LowDegreeBusInteractionOptimizer<'a, T, V, S, B> {
```

Comment:
As long as the struct has no real state, I found it easier to use free functions with additional parameters instead, especially wit generics involved. In this case, I think that many functions only need one of the three fields here, so you don't even have more parameters.